### PR TITLE
Add cross-regional StorageClass support

### PR DESCRIPTION
**Pull request by Prachi030510**
## Checklist before requesting a review
- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] I have made sure existing UTs are not impacted
- [ ] I have executed all necessary linter and made sure internal pipeline/travis is not broken

### CODE REVIEW

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

**Comment by Prachi030510** [comment]
```bash
% kubectl get cm cluster-info -n kube-system -o yaml                                                                                  
apiVersion: v1
data:
  account_id: 68010fd8df4f467681ddec1e065d7a48
  cluster-config.json: |
    {
      "cluster_id": "d7s3na2s093gq0qsdm20",
      "cluster_name": "prod-syd01-d7s3na2s093gq0qsdm20",
      "cluster_type": "vpc-gen2_cruiser",
      "cluster_pay_tier": "paid",
      "datacenter": "syd01",
      "account_id": "68010fd8df4f467681ddec1e065d7a48",
      "created": "2026-05-04T06:25:41+0000",
      "name": "psa-ap-cluster",
      "offering": "openshift",
      "master_public_url": "https://c100-e.au-syd.containers.cloud.ibm.com:31254",
      "master_url": "https://c100.au-syd.containers.cloud.ibm.com:31254",
      "crn":  "crn:v1:bluemix:public:containers-kubernetes:au-syd:68010fd8df4f467681ddec1e065d7a48:d7s3na2s093gq0qsdm20"
    }
  cluster_id: d7s3na2s093gq0qsdm20
  cluster_name: prod-syd01-d7s3na2s093gq0qsdm20
  cluster_pay_tier: paid
  cluster_type: vpc-gen2_cruiser
  created: 2026-05-04T06:25:41+0000
  crn: crn:v1:bluemix:public:containers-kubernetes:au-syd:68010fd8df4f467681ddec1e065d7a48:d7s3na2s093gq0qsdm20
  datacenter: syd01
  etcd_host: c100.au-syd.containers.cloud.ibm.com
  kube_1_33: 4.20.21_1548_openshift
  master_public_url: https://c100-e.au-syd.containers.cloud.ibm.com:31254
  master_url: https://c100.au-syd.containers.cloud.ibm.com:31254
  name: psa-ap-cluster
  offering: openshift
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: '{"apiVersion":"v1","data":{"account_id":"68010fd8df4f467681ddec1e065d7a48","cluster-config.json":"{\n  \"cluster_id\":
      \"d7s3na2s093gq0qsdm20\",\n  \"cluster_name\": \"prod-syd01-d7s3na2s093gq0qsdm20\",\n  \"cluster_type\":
      \"vpc-gen2_cruiser\",\n  \"cluster_pay_tier\": \"paid\",\n  \"datacenter\":
      \"syd01\",\n  \"account_id\": \"68010fd8df4f467681ddec1e065d7a48\",\n  \"created\":
      \"2026-05-04T06:25:41+0000\",\n  \"name\": \"psa-ap-cluster\",\n  \"offering\":
      \"openshift\",\n  \"master_public_url\": \"https://c100-e.au-syd.containers.cloud.ibm.com:31254\",\n  \"master_url\":
      \"https://c100.au-syd.containers.cloud.ibm.com:31254\",\n  \"crn\":  \"crn:v1:bluemix:public:containers-kubernetes:au-syd:68010fd8df4f467681ddec1e065d7a48:d7s3na2s093gq0qsdm20\"\n}\n","cluster_id":"d7s3na2s093gq0qsdm20","cluster_name":"prod-syd01-d7s3na2s093gq0qsdm20","cluster_pay_tier":"paid","cluster_type":"vpc-gen2_cruiser","created":"2026-05-04T06:25:41+0000","crn":"crn:v1:bluemix:public:containers-kubernetes:au-syd:68010fd8df4f467681ddec1e065d7a48:d7s3na2s093gq0qsdm20","datacenter":"syd01","etcd_host":"c100.au-syd.containers.cloud.ibm.com","kube_1_33":"4.20.21_1548_openshift","master_public_url":"https://c100-e.au-syd.containers.cloud.ibm.com:31254","master_url":"https://c100.au-syd.containers.cloud.ibm.com:31254","name":"psa-ap-cluster","offering":"openshift"},"kind":"ConfigMap","metadata":{"name":"cluster-info","namespace":"kube-system"}}'
  creationTimestamp: "2026-05-04T06:39:53Z"
  name: cluster-info
  namespace: kube-system
  resourceVersion: "4746494"
  uid: d2fa35d8-5bfe-48be-b5d9-4ea15dad2cd3
```
```bash
%  kubectl get sc | grep ibm-object                
ibm-object-storage-smart-cross-region-rclone             cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-smart-cross-region-rclone-retain      cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-smart-cross-region-s3fs               cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-smart-cross-region-s3fs-retain        cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-smart-rclone                          cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-smart-rclone-retain                   cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-smart-s3fs                            cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-smart-s3fs-retain                     cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-standard-cross-region-rclone          cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-standard-cross-region-rclone-retain   cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-standard-cross-region-s3fs            cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-standard-cross-region-s3fs-retain     cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-standard-rclone                       cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-standard-rclone-retain                cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s
ibm-object-storage-standard-s3fs                         cos.s3.csi.ibm.io      Delete          Immediate              false                  2m11s
ibm-object-storage-standard-s3fs-retain                  cos.s3.csi.ibm.io      Retain          Immediate              false                  2m11s

%  kubectl describe sc ibm-object-storage-smart-cross-region-rclone ibm-object-storage-standard-cross-region-s3fs
Name:                  ibm-object-storage-smart-cross-region-rclone
IsDefaultClass:        No
Annotations:           <none>
Provisioner:           cos.s3.csi.ibm.io
Parameters:            client=awss3,cosEndpoint=https://s3.direct.ap.cloud-object-storage.appdomain.cloud,csi.storage.k8s.io/node-publish-secret-name=${pvc.annotations['cos.csi.driver/secret']},csi.storage.k8s.io/node-publish-secret-namespace=${pvc.namespace},locationConstraint=ap-smart,mounter=rclone
AllowVolumeExpansion:  <unset>
MountOptions:
  acl=private
  bucket_acl=private
  upload_cutoff=100Mi
  chunk_size=16Mi
  max_upload_parts=1000
  upload_concurrency=8
  multi_thread_streams=8
  disable_checksum=true
ReclaimPolicy:      Delete
VolumeBindingMode:  Immediate
Events:             <none>


Name:                  ibm-object-storage-standard-cross-region-s3fs
IsDefaultClass:        No
Annotations:           <none>
Provisioner:           cos.s3.csi.ibm.io
Parameters:            client=awss3,cosEndpoint=https://s3.direct.ap.cloud-object-storage.appdomain.cloud,csi.storage.k8s.io/node-publish-secret-name=${pvc.annotations['cos.csi.driver/secret']},csi.storage.k8s.io/node-publish-secret-namespace=${pvc.namespace},locationConstraint=ap-standard,mounter=s3fs
AllowVolumeExpansion:  <unset>
MountOptions:
  multipart_size=52
  multireq_max=20
  max_dirty_data=5120
  parallel_count=20
  max_stat_cache_size=100000
  retries=5
  kernel_cache
ReclaimPolicy:      Delete
VolumeBindingMode:  Immediate
Events:             <none>

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

File: controllers/constants/constants.go [hunk 1]
```diff
@@ -97,10 +97,31 @@ const (
 	IaasIBMClassic   = "ibm-classic"
 	IaasIBMVPC       = "ibm-vpc"
 
-	IBMEP    = "https://s3.%s.%s.cloud-object-storage.appdomain.cloud"
-	AWSEP    = "https://s3.%s.amazonaws.com"
-	WasabiEP = "https://s3.%s.wasabisys.com"
+	IBMEP         = "https://s3.%s.%s.cloud-object-storage.appdomain.cloud"
+	IBMCrossRegEP = "https://s3.%s.%s.cloud-object-storage.appdomain.cloud"
```
Comment (mssachan) [inline controllers/constants/constants.go]: As discussed over call, since cos endpoint format `"https://s3.%s.%s.cloud-object-storage.appdomain.cloud"` is same for both regiona and cross-regional endpoint, define only one constant.

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

File: controllers/constants/constants.go [hunk 2]
```diff
@@ -97,10 +97,31 @@ const (
...
+	"eu-es":    "eu",
+	"eu-fr2":   "eu",
+	"au-syd":   "ap",
+	"jp-tok":   "ap",
+	"jp-osa":   "ap",
+	"in-che":   "ap",
+	"in-mum":   "ap",
+}
 
+const (
```
Comment (mssachan) [inline controllers/constants/constants.go]: why 2 different `const` sections? Move `var RegionToGeography = map[string]string{}` after `consts` section.

File: controllers/internal/crutils/static_resource_generator.go [hunk 3]
```diff
@@ -228,13 +228,22 @@ func (c *IBMObjectCSI) GenerateSCCForNodeClusterRoleBinding() *rbacv1.ClusterRol
 func (c *IBMObjectCSI) GenerateS3fsSC(scInputParams SCInputParams) *storagev1.StorageClass {
 	var storageClassName, locationConstraint string
 	if scInputParams.S3Provider == constants.S3ProviderIBM {
-		locationConstraint = fmt.Sprintf("%s-%s", scInputParams.Region, scInputParams.COSStorageClass)
+		if scInputParams.IsCrossRegional {
+			geography := constants.RegionToGeography[scInputParams.Region]
+			locationConstraint = fmt.Sprintf("%s-%s", geography, scInputParams.COSStorageClass)
```
Comment (mssachan) [inline controllers/internal/crutils/static_resource_generator.go]: what happens when cos-csi addon is enabled on a cluster in the NEW region which is still not supported by our cos-csi driver?

File: controllers/internal/crutils/static_resource_generator.go [hunk 4]
```diff
@@ -272,14 +281,23 @@ func (c *IBMObjectCSI) GenerateS3fsSC(scInputParams SCInputParams) *storagev1.St
...
+	}
 	if scInputParams.ReclaimPolicy == corev1.PersistentVolumeReclaimRetain {
 		storageClassName = fmt.Sprintf("%s-%s", storageClassName, constants.RetainPolicyTag) // "ibm-object-storage-standard-rclone-retain"
 	}
 
 	if scInputParams.S3Provider == constants.S3ProviderIBM {
-		locationConstraint = fmt.Sprintf("%s-%s", scInputParams.Region, scInputParams.COSStorageClass)
+		if scInputParams.IsCrossRegional {
+			geography := constants.RegionToGeography[scInputParams.Region]
+			locationConstraint = fmt.Sprintf("%s-%s", geography, scInputParams.COSStorageClass)
```
Comment (mssachan) [inline controllers/internal/crutils/static_resource_generator.go]: same here

File: controllers/ibmobjectcsi_controller.go [hunk 5]
```diff
@@ -583,13 +583,67 @@ func (r *IBMObjectCSIReconciler) getStorageClasses(instance *crutils.IBMObjectCS
...
+						S3Provider:      s3Provider,
+						Region:          requiredRegion,
+						COSEndpoint:     crossRegCosEP,
+						COSStorageClass: sc,
+						IsCrossRegional: true,
+					})
+					k8sSCs = append(k8sSCs, s3fsK8sSc)
+				}
+			}
+		}
```
Comment (mssachan) [inline controllers/ibmobjectcsi_controller.go:607]: I think you need to add `return k8sSCs` here after 640 for provider `IBM`

File: controllers/util/common/common.go [hunk 6]
```diff
@@ -428,13 +428,76 @@ func (ch *ControllerHelper) GetIBMCosSC() []string {
...
+			region = "NA"
+		}
+	}
+	
+	if ch.IaaSProvider == constants.IaasIBMVPC || ch.IaaSProvider == constants.IaasIBMClassic {
+		epType := "private"
+		if ch.IaaSProvider == constants.IaasIBMVPC {
+			epType = "direct"
+		}
+		ch.CosEP = fmt.Sprintf(constants.IBMEP, epType, region)
```
Comment (mssachan) [inline controllers/util/common/common.go]: in case of new region, `region` will be `NA` and
cos endpoint will become `https://s3.direct.NA.xxxxxx`

File: controllers/util/common/common.go [hunk 7]
```diff
@@ -428,13 +428,76 @@ func (ch *ControllerHelper) GetIBMCosSC() []string {
...
+
+	geography, supported := constants.RegionToGeography[ch.Region]
+	if !supported {
+		// For unknown regions, distinguish between preprod and future prod
+		if isPreprodRegion(ch.Region) {
+			// Preprod regions (dev/stage/prestage) use US geography
+			geography = "us"
+		} else {
+			// Future prod regions not yet in map use NA.
+			geography = "na"
```
Comment (mssachan) [inline controllers/util/common/common.go]: use consisten value `NA` instead of `na`

File: controllers/util/common/common.go [hunk 8]
```diff
@@ -428,13 +428,76 @@ func (ch *ControllerHelper) GetIBMCosSC() []string {
...
+		}
+	}
+
 	if ch.IaaSProvider == constants.IaasIBMVPC || ch.IaaSProvider == constants.IaasIBMClassic {
 		epType := "private"
 		if ch.IaaSProvider == constants.IaasIBMVPC {
 			epType = "direct"
 		}
-		ch.CosEP = fmt.Sprintf(constants.IBMEP, epType, ch.Region)
+		ch.CosEP = fmt.Sprintf(constants.IBMCrossRegEP, epType, geography)
```
Comment (mssachan) [inline controllers/util/common/common.go]: same here

File: controllers/util/common/common.go [hunk 9]
```diff
@@ -428,13 +428,76 @@ func (ch *ControllerHelper) GetIBMCosSC() []string {
...
+	if ch.IaaSProvider == constants.IaasIBMVPC || ch.IaaSProvider == constants.IaasIBMClassic {
+		epType := "private"
+		if ch.IaaSProvider == constants.IaasIBMVPC {
+			epType = "direct"
+		}
+		ch.CosEP = fmt.Sprintf(constants.IBMEP, epType, region)
+	}
+}
+// isPreprodRegion checks if region is preprod (dev/stage/prestage/test)
+func isPreprodRegion(region string) bool {
```
Comment (mssachan) [inline controllers/util/common/common.go]: Revisit logic to check preprod clusters

**Review by mssachan — CHANGES_REQUESTED** [review]
Added comments in PR.

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

File: controllers/internal/crutils/static_resource_generator.go [hunk 10]
```diff
@@ -229,7 +229,10 @@ func (c *IBMObjectCSI) GenerateS3fsSC(scInputParams SCInputParams) *storagev1.St
 	var storageClassName, locationConstraint string
 	if scInputParams.S3Provider == constants.S3ProviderIBM {
 		if scInputParams.IsCrossRegional {
-			geography := constants.RegionToGeography[scInputParams.Region]
+			geography, exists := constants.RegionToGeography[scInputParams.Region]
+			if !exists {
+				geography = "NA"
```
Comment (mssachan) [inline controllers/internal/crutils/static_resource_generator.go]: still with this change, now locationConstraint will become `NA-xyz`.
What you need to do here is entire `locationConstraint` should be `NA` incase region is not recognised.

File: controllers/internal/crutils/static_resource_generator.go [hunk 11]
```diff
@@ -293,7 +296,10 @@ func (c *IBMObjectCSI) GenerateRcloneSC(scInputParams SCInputParams) *storagev1.
 
 	if scInputParams.S3Provider == constants.S3ProviderIBM {
 		if scInputParams.IsCrossRegional {
-			geography := constants.RegionToGeography[scInputParams.Region]
+			geography, exists := constants.RegionToGeography[scInputParams.Region]
+			if !exists {
+				geography = "NA"
```
Comment (mssachan) [inline controllers/internal/crutils/static_resource_generator.go]: same here

File: controllers/util/common/common.go [hunk 12]
```diff
@@ -449,30 +445,6 @@ func (ch *ControllerHelper) SetIBMCosEP() {
 		ch.CosEP = fmt.Sprintf(constants.IBMEP, epType, region)
 	}
 }
-// isPreprodRegion checks if region is preprod (dev/stage/prestage/test)
-func isPreprodRegion(region string) bool {
```
Comment (mssachan) [inline controllers/util/common/common.go]: did u open new GHE issue to deliver this in follow-up PR?

**Review by mssachan — CHANGES_REQUESTED** [review]
Added minor comment. PTAL

**Comment by mssachan** [comment]
@Prachi030510 Also add the test results for negative scenario (I mean simulate the scenario of Newly added region which is NOT supported by our addon yet)

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

File: controllers/util/common/common.go (hunk 12 above)
Comment (Prachi030510) [reply controllers/util/common/common.go]: GHE issue - https://github.ibm.com/alchemy-containers/armada-storage/issues/9620

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.51%

**Comment by contsto** [comment]
:red_circle: Coverage decreased from 21.64% to 21.56%

**Review by mssachan — APPROVED** [review]
LGTM
